### PR TITLE
Fix stuck OpenHands terminal log

### PR DESCRIPTION
**Pull request by enyst**
Fixes the integrated `OpenHands` terminal view getting stuck / not updating after the first command.

- Replace `Terminal.sendText()` output replay (which can corrupt the shell/session) with a dedicated `vscode.Pseudoterminal` log.
- Normalize newlines + chunk large writes to avoid renderer issues.
- Add a unit test that emits `BashCommand`/`BashOutput` and asserts the PTY receives the output.

Tested: `npm test`

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive tests for local terminal behavior: recreation after close, ANSI/emoji across chunk boundaries, streaming multiple Bash events, command/error handling, newline normalization, large-output chunking, and terminal-close mocking.

* **Improvements**
  * Route terminal I/O through a read-only terminal layer for more reliable logging and lifecycle handling.
  * Improve output chunking and newline normalization, per-command exit summaries, synthesized process-exit footers, sensitive-value redaction, and cleanup on close/config changes.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by enyst** [comment]
@openhands Read this PR and do a /codereview-roasted on it.

**Comment by enyst** [comment]
Taste Rating: 🔴 Needs improvement

Linus-Style Analysis

[CRITICAL ISSUES]
- [src/extension.ts, Lines 242–254, 98–105] Correctness: Terminal log dies after user closes it and never recovers. You cache terminal and pty in globals and only recreate on “!terminal || !terminalLogPty”. When the user closes the terminal, VS Code will call pty.close(), setting your internal closed flag, but your “terminal” reference remains non-null. Subsequent events hit write(), which no-ops because closed === true, and you never recreate the terminal since “terminal” is truthy. Fix by listening to vscode.window.onDidCloseTerminal and/or wire pty.onDidClose to clear both terminal and terminalLogPty references so the next event recreates it. Also guard on terminalLogPty.closed.
- [src/extension.ts, Lines 118–123] Data Integrity: Naive chunking can split ANSI escape sequences or UTF-16 surrogate pairs. Your 16k slicing can bisect CSI or emoji, causing garbled output or broken color states. Use a chunker that respects ANSI boundaries and code points (track ESC[..m sequences) or at minimum split on newline boundaries and keep a small carryover buffer when the chunk ends within an ESC sequence.
- [src/extension.ts, Lines 256–264] Robustness: Exit code is only shown on BashExit events. BashOutput may carry exit_code (your own test sets it) but you ignore it. If BashExit never arrives (interruption, emitter bug), users never see the exit summary. Either synthesize a footer when you see exit_code on output or ensure the emitter always delivers BashExit and add a defensive fallback.
- [src/extension.ts, Lines 298–316] Lifecycle: Mode switch and reconnection don’t dispose the log terminal. Switching from local→remote leaves an orphaned read‑only terminal that continues to occupy UI space. This is part of the stuck/dead-terminal UX. Dispose the “OpenHands” log terminal when leaving local mode and recreate when re-entering.

[IMPROVEMENT OPPORTUNITIES]
- [src/extension.ts, Lines 83, 115–123] Newline handling: normalizeTerminalNewlines only addresses LF→CRLF. Many CLIs use bare “\r” carriage returns for progress bars. Today those CRs pass through but can combine poorly with your unconditional LF→CRLF conversion. Consider handling CR-only sequences: coalesce “...text\rtext2” carriage-return updates and avoid adding spurious blank lines.
- [src/extension.ts, Lines 242–265] UX signal: Writing a leading blank line before every command (writeLine('')) adds vertical bloat. Make it conditional (e.g., only if previous write wasn’t a newline) or add a thin delimiter once per tool run.
- [src/extension.ts, Lines 136–153, 323–360] Logging hygiene: safeStringify might leak secrets if events contain tokens/keys. You already special-case LLM stream logs; extend that to redact known secret-like fields (api_key, authorization) from all logged events before appending to the OutputChannel.
- [src/__tests__/extension.test.ts, Lines 323–361] Tests under-specified: No coverage for terminal close/recreate, ANSI splitting, CR-only progress lines, or stderr interleaving. Add tests for: user closes terminal then another BashEvent arrives (should recreate); long ANSI-colored line that crosses chunk boundary; repetitive “\r” updates; stderr writing in a different color prefix if desired.

[STYLE NOTES]
- [src/extension.ts, Lines 230–268] Single-responsibility: handleTerminalEvent is doing routing, lifecycle (create terminal), and rendering. Split into: ensureTerminal(), and renderTerminalEvent().
- [src/extension.ts, Lines 270–404] ensurePanelAndConnection is bloated and mixes UI setup, settings, conversation lifecycle, and event wire-up. Extract helpers to reduce nesting and make failure cases clearer.

Verdict: ❌ Needs rework
Solid direction—moving to a pseudoterminal is the right fix to avoid corrupting a real shell—but the lifecycle hole means the log silently dies after close, and the chunking risks corrupting terminal output. Address those, and this becomes mergeable.

Key Insight:
You fixed the symptom (stuck log) by swapping transport, but didn’t harden the lifecycle or data-path: if the terminal closes or output crosses chunk boundaries, the log quietly fails or garbles—solve those two and the design holds.

**Comment by enyst** [comment]
@openhands Read all review threads and comments on this PR, including top level ones, and including those on source files/lines. Investigate, understand them, and if you agree they’re issues, fix them, and for those you don’t agree, that’s okay. Commit and push. Then comment on the PR with your fixes and explanations if you didn’t want to address something.

IMPORTANT: you are an agent, execute the entire task and do not ask me anything. Read all and perform the full task before giving any message for the user.

File: src/extension.ts [hunk 1]
```diff
@@ -79,6 +80,53 @@ function fileLog(line: string) {
...
+    this.writeLine('[OpenHands] This terminal is read-only.');
+    this.writeLine('[OpenHands] Use a normal VS Code terminal for manual commands.');
+    this.writeLine('');
+  }
+
+  write(text: string): void {
+    if (this.closed) return;
+    const normalized = normalizeTerminalNewlines(text);
+    // Avoid very large single writes which can fail or lag the terminal renderer.
+    const chunkSize = 16_000;
```
[the remark this answers is not in the document]
Comment (enyst) [reply src/extension.ts]: /gemini wdyt now?

File: src/__tests__/extension.test.ts [hunk 2]
```diff
@@ -319,6 +319,46 @@ describe('Settings and modes', () => {
...
+      order: 1,
+      exit_code: 0,
+      stdout: '/test/workspace\n',
+      stderr: null,
+    });
+
+    expect(vscode.window.createTerminal).toHaveBeenCalledWith(expect.objectContaining({ name: 'OpenHands' }));
+    expect(writes.join('')).toContain('$ pwd && ls -la');
+    expect(writes.join('')).toContain('/test/workspace');
+  });
```
[the remark this answers is not in the document]
Comment (enyst) [reply src/__tests__/extension.test.ts:519]: /gemini wdyt now?

File: src/extension.ts [hunk 3]
```diff
@@ -191,25 +239,28 @@ export function activate(context: vscode.ExtensionContext) {
...
+        terminal = vscode.window.createTerminal({ name: 'OpenHands', pty: terminalLogPty });
         terminal.show(true);
       } catch (e) {
-        console.error('[Terminal] Failed to create terminal:', e);
+        console.error('[Terminal] Failed to create terminal log:', e);
+        terminal = undefined;
+        terminalLogPty = undefined;
         return;
       }
     }
```
[the remark this answers is not in the document]
Comment (enyst) [reply src/extension.ts]: /gemini wdyt now?

**Comment by enyst** [comment]
@OpenHands please fix the failing actions on PR #196 at branch `fix/terminal-log-view`

Read the PR comments and re-verify the reviews, are all issues solved? If not, solve them please.

You are an agent, continue working until you are completely done.

**Comment by enyst** [comment]
I fixed the failing actions and re-verified review feedback:

- Fixed unit test assertion for newline normalization in the OpenHands PTY log. The PTY intentionally normalizes newlines to CRLF; updated the test to look for CRLF in the echoed command string. This aligns with the existing normalizeTerminalNewlines() behavior and avoids false negatives.
- Re-ran lint, build, and full test suites locally — all green.
- Pushed commit d03402d with the test fix.
- CI workflows have been re-triggered and are in progress for this head SHA.

Review threads status:
- Type-safety for command_id/exit_code and BashExit handling — addressed previously (commit b3a5443), confirmed by CodeRabbit thread resolution.
- Terminal lifecycle (onDidCloseTerminal handling and PTY recreation) — addressed in earlier commits (3ad1e29), and verified.

I will monitor the GitHub Actions runs for this PR to ensure they complete successfully. If anything else surfaces, I’ll follow up.

**Comment by enyst** [comment]
I fixed the failing actions and re-verified review feedback:

- Fixed unit test assertion for newline normalization in the OpenHands PTY log. The PTY intentionally normalizes newlines to CRLF; updated the test to look for CRLF in the echoed command string. This aligns with the existing normalizeTerminalNewlines() behavior and avoids false negatives.
- Re-ran lint, build, and full test suites locally — all green.
- Pushed commit d03402d with the test fix.
- CI workflows have been re-triggered and are in progress for this head SHA.

Review threads status:
- Type-safety for command_id/exit_code and BashExit handling — addressed previously (commit b3a5443), confirmed by CodeRabbit thread resolution.
- Terminal lifecycle (onDidCloseTerminal handling and PTY recreation) — addressed in earlier commits (3ad1e29), and verified.

I will monitor the GitHub Actions runs for this PR to ensure they complete successfully. If anything else surfaces, I’ll follow up.

**Comment by enyst** [comment]
Addressed AI review suggestions after verifying correctness:

1) Add tests for terminal close/recreate lifecycle — ACCEPTED and implemented
- Verified the extension listens to onDidCloseTerminal and clears cached references
- Added unit test "recreates the terminal after user closes it" to ensure the PTY/terminal is recreated and subsequent commands log correctly

2) Test large output with ANSI escape sequences and Unicode across chunk boundary — ACCEPTED and implemented
- Added unit test "handles ANSI sequences and emoji across chunk boundary"
- Confirms chunking preserves ANSI CSI (ESC[ … terminator) and does not split surrogate pairs (e.g., 🚀) while normalizing newlines to CRLF

Rationale: Both align with the PR’s stated goals (robust PTY lifecycle and chunking safety). Local lint/build/tests pass; CI will validate on this head.

Commits:
- d711204 tests: add coverage for terminal close/recreate and ANSI+emoji chunking

Let me know if you want additional edge cases covered.
